Readme.pod
=pod

=head1 NAME

Mojolicious::Plugin::Util::Endpoint - Use Template URIs in Mojolicious


=head1 SYNOPSIS

  # Mojolicious::Lite
  plugin 'Util::Endpoint';

  # Mojolicious
  $self->plugin('Util::Endpoint');

  my $rs = $mojo->routes;

  # Set endpoint
  my $r = $rs->route('/:user')->endpoint(
    webfinger => {
      query  => [
        q => '{uri}'
      ]
    });

  return $self->endpoint('webfinger');
  # https://sojolicio.us/{user}?q={uri}

  $self->stash(user => 'Akron');

  return $self->endpoint('webfinger');
  # https://sojolicio.us/Akron?q={uri}

  return $self->endpoint(webfinger => {
    uri => 'acct:akron@sojolicio.us'
  });
  # https://sojolicio.us/Akron?q=acct%3Aakron%40sojolicio.us


=head1 DESCRIPTION

L<Mojolicious::Plugin::Util::Endpoint> is a plugin that
allows for the simple establishment of endpoint URIs.
This is similar to L<url_for|Mojolicious::Controller/url_for>,
but includes support for template URIs with parameters
following L<RFC6570|https://tools.ietf.org/html/rfc6570> Level 1
(as used in, e.g., Host-Meta or OpenSearch).


=head1 METHODS

=head2 register

  # Mojolicious
  $app->plugin('Util::Endpoint');

  # Mojolicious::Lite
  plugin 'Util::Endpoint';

Called when registering the plugin.


=head1 SHORTCUTS

=head2 endpoint

  my $rs = $mojo->routes
  my $r = $rs->route('/suggest')->endpoint(
    opensearch => {
      scheme => 'https',
      host   => 'sojolicio.us',
      port   => 3000,
      query  => [
        q     => '{searchTerms}',
        start => '{startIndex?}'
      ]
    });

Establishes an endpoint defined for a service.
It accepts optional parameters C<scheme>, C<host>,
a C<port> and query parameters (C<query>),
overwriting the current values of C<url_for>.
Template parameters need curly brackets, optional
template parameters need a question mark before
the closing bracket.
Optional path placeholders are currenty not supported.
Returns the route.

B<Warning>: Support for named routes to use with
C<url_for> was dropped in v0.19.

=head1 HELPERS

=head2 endpoint

  # In Controller:
  #   Set endpoints:
  $self->endpoint(hub => 'http://sojolicio.us/search?q={searchTerm}');
  $self->endpoint(hub => Mojo::URL->new('http://pubsubhubbub.appspot.com/'));

  #   Get endpoints:
  return $self->endpoint('webfinger');
  return $self->endpoint(webfinger => { user => 'me' } );

  # Interpolate arbitrary template URIs
  return $self->endpoint(
    'http://sojolicio.us/.well-known/webfinger?resource={uri}&rel={rel?}' => {
      'uri' => 'acct:akron@sojolicio.us',
      '?'   => undef
    });

  # In Template:
  <%= endpoint 'webfinger' %>

Get or set endpoints defined for a specific service.

For setting it accepts the name of the endpoint and
either a string with the endpoint URI or a L<Mojo::URL> object.

For getting it accepts the name of the endpoint or an arbitrary
template URI and additional stash values for the route as a hash reference.
These stash values override existing stash values from
the controller and fill the template variables.

  # In Controller:
  return $self->endpoint('opensearch');
  # https://sojolicio.us/suggest?q={searchTerms}&start={startIndex?}

  return $self->endpoint(opensearch => {
    searchTerms => 'simpson',
    '?' => undef
  });
  # https://sojolicio.us/suggest?q=simpson

The special parameter C<?> can be set to C<undef> to ignore
all undefined optional template parameters.

If the defined endpoint can't be found, the value for C<url_for>
is returned.


=head2 get_endpoints

  # In Controller:
  my $hash = $self->get_endpoints;

  while (my ($key, $value) = each %$hash) {
    print $key, ' => ', $value, "\n";
  };

Returns a hash of all endpoints, interpolated with the current
controller stash.

B<Note:> This helper is EXPERIMENTAL and may be deprecated in further releases.


=head1 COMMANDS

=head2 endpoints

  $ perl app.pl endpoints

Show all endpoints of the app established by this plugin.


=head1 DEPENDENCIES

L<Mojolicious> (best with SSL support).


=head1 CONTRIBUTORS

L<Viacheslav Tykhanovskyi|https://github.com/vti>


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Util-Endpoint


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011-2016, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the same terms as Perl.

=cut


Changes
0.20 2016-09-13
        - Improved test coverage.
        - Improved access rights to be Kwalitee compliant.

0.19 2016-02-29
        - Fixed for Mojolicious 6.48
          (This removes support for named routes usable for url_for).

0.18 2014-04-18
        - Fixed dependency.

0.17 2014-04-14
        - Fixed compatibility with Mojolicious 6.0 (vti).
	  This was a contribution motivated by the
          CPAN Pull Request Challenge - Thank you very much!

0.16 2014-02-10
        - Documentation tweaks regarding the template URI RFC.

0.15 2013-05-20
        - Fixed compatibility with Mojolicious 4.0 (Reneeb).
	- Removed experimental given/when operations (Reneeb, Akron).
	- Fixed randomized hash bug in t/Endpoint-Command.t.

0.14 2013-04-17
        - Improve documentation.
	- Fixed bug with named routes.

0.13 2013-04-07
        - Allow interpolation of arbitrary urls.
        - Small improvement to documentation.
	- Support undefining placeholders in path, domain and parameters.
	- Small performance improvement for '?' => undef.
	- Improved test suite.

0.12 2013-03-22
        - Made warning for undefined endpoints a
          debug information.

0.11 2013-03-18
        - Fixed a newly introduced relative path bug.

0.10 2013-03-18
        - Partially reimplemented due to
	  a bug with mounted apps.
	- Update dependencies.

0.09 2013-03-16
        - Update year.
	- Documentation tweaks.
	- Change title of endpoints command.
	- Small bugfix in port setting.

0.08 2012-11-19
        - Small bugfix regarding empty queries.

0.07 2012-11-13
        - Small bugfix regarding warnings.

0.06 2012-11-03
        - Small improvement regarding no-placeholder URIs.
        - Small improvements of the documentation.

0.05 2012-10-30
        - Add setting function to endpoint helper.
	- Small improvements of the test suite.

0.04 2012-10-28
        - Small improvements/corrections of the documentation.

0.03 2012-10-28
        - Corrections regarding meta data.
	- Small improvements of the documentation.

0.02 2012-10-28
        - Small improvements/corrections of the documentation.
	- Small code improvements.

0.01 2012-10-28
        - Separation from Sojolicious package.
        - Initial submission to GitHub.
        - Initial submission to CPAN.


t/endpoint-mounted.pl
#!/usr/bin/env perl
use Mojolicious::Lite;

use lib '../lib';

plugin 'Util::Endpoint';

get '/test' => sub {
  shift->render_text('Mounted.');
};

(get '/probe')->to(
  cb => sub {
    shift->render_text('Mounted Endpoint.')
  })->endpoint('probe');

get '/get-ep' => sub {
  my $c = shift;
  return $c->render_text($c->endpoint('probe'));
};

get '/get-url' => sub {
  my $c = shift;
  return $c->render_text($c->url_for('probe'));
};

app->start;


lib/Mojolicious/Plugin/Util/Endpoint/endpoints.pm
package Mojolicious::Plugin::Util::Endpoint::endpoints;
use Mojo::Base 'Mojolicious::Command';

use Getopt::Long qw/GetOptions :config no_auto_abbrev no_ignore_case/;

has description => "Show available endpoints.\n";
has usage       => <<"EOF";
usage: $0 endpoints

  perl app.pl endpoints

EOF


# Run endpoints
sub run {
  my $self = shift;

  # Options
  local @ARGV = @_;

  my $c = Mojolicious::Controller->new;
  $c->app($self->app);

  # Get endpoints
  my $endpoints = $c->get_endpoints;

  # No endpoints
  return unless $endpoints;

  # Print all endpoints
  foreach my $name (sort { $a cmp $b } keys %$endpoints) {
    printf " %-20s %s\n", qq{"$name"}, $endpoints->{$name};
  };
  print "\n";

  return;
};


1;


__END__

=pod

=head1 NAME

Mojolicious::Plugin::Util::Endpoint::endpoints - List Template URIs

=head1 SYNOPSIS

  use Mojolicious::Plugin::Util::Endpoint::endpoints;

  my $ep = Mojolicious::Plugin::Util::Endpoint::endpoints->new;
  $ep->run;


=head1 DESCRIPTION

L<Mojolicious::Plugin::Util::Endpoint::endpoints> lists all
endpoints established by L<Mojolicious::Plugin::Util::Endpoint>.

It is normally started from the command line:

  $ perl app.pl endpoints


=head1 ATTRIBUTES

L<Mojolicious::Plugin::Util::Endpoint::endpoints> inherits all
attributes from L<Mojolicious::Command> and implements the following
new ones.


=head2 description

  my $description = $ep->description;
  $ep = $ep->description('Foo!');

Short description of this command, used for the command list.


=head2 usage

  my $usage = $ep->usage;
  $ep       = $ep->usage('Foo!');

Usage information for this command, used for the help screen.


=head1 METHODS

L<Mojolicious::Plugin::Util::Endpoint::endpoints> inherits all
methods from L<Mojolicious::Command> and implements the following new ones.


=head2 run

  $ep->run;

Run this command.


=head1 DEPENDENCIES

L<Mojolicious>,
L<Mojolicious::Plugin::Util::Endpoint>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Util-Endpoint


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012-2014, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the same terms as Perl.

The documentation is based on L<Mojolicious::Command::eval>,
written by Sebastian Riedel.

=cut


lib/Mojolicious/Plugin/Util/Endpoint.pm
package Mojolicious::Plugin::Util::Endpoint;
use Mojo::Base 'Mojolicious::Plugin';
use Mojo::ByteStream 'b';
use Scalar::Util qw/blessed/;
use Mojo::URL;

our $VERSION = '0.20';

# Todo: Support alternative bases for https-paths
# Todo: Update to https://tools.ietf.org/html/rfc6570
# Todo: Allow for changing scheme, port, host etc. afterwards

# Endpoint hash
our %endpoints;

# Register Plugin
sub register {
  my ($plugin, $mojo) = @_;

  # Add 'endpoints' command
  push @{$mojo->commands->namespaces}, __PACKAGE__;

  # Add 'endpoint' shortcut
  $mojo->routes->add_shortcut(
    endpoint => sub {
      my ($route, $name, $param) = @_;

      # Endpoint already defined
      if (exists $endpoints{$name}) {
        $mojo->log->debug(qq{Route endpoint "$name" already defined});
        return $route;
      };

      $param //= {};

      # Route defined
      $param->{route} = $route->name;

      # Set to stash
      $endpoints{$name} = $param // {};

      # Return route for piping
      return $route;
    }
  );


  # Add 'endpoint' helper
  $mojo->helper(
    endpoint => sub {
      my ($c, $name, $values) = @_;
      $values ||= {};

      # Define endpoint by string
      unless (ref $values) {
        return ($endpoints{$name} = Mojo::URL->new($values));
      }

      # Define endpoint by Mojo::URL
      elsif (blessed $values && $values->isa('Mojo::URL')) {
        return ($endpoints{$name} = $values->clone);
      };

      # Set values
      my %values = (
        $c->isa('Mojolicious::Controller') ? %{$c->stash} : %{$c->defaults},
        format => undef,
        %$values
      );

      # Endpoint undefined
      unless (defined $endpoints{$name}) {

        # Named route
        if ($name !~ m!^([^:]+:)?/?/!) {
          return $c->url_for($name)->to_abs->to_string;
        };

        # Interpolate string
        return _interpolate($name, \%values, $values);
      };

      # Return interpolated string
      if (blessed $endpoints{$name} && $endpoints{$name}->isa('Mojo::URL')) {
        return _interpolate(
          $endpoints{$name}->to_abs->to_string,
          \%values,
          $values
        );
      };

      # The following is based on url_for of Mojolicious::Controller
      # and parts of path_for in Mojolicious::Routes::Route
      # Get match object
      my $match;
      unless ($match = $c->match) {
        $match = Mojolicious::Routes::Match->new(get => '/');
        $match->root($c->app->routes);
      };

      # Base
      my $url = Mojo::URL->new;
      my $req = $c->req;
      $url->base($req->url->base->clone);
      my $base = $url->base;
      $base->userinfo(undef);

      # Get parameters
      my $param = $endpoints{$name};

      # Set parameters to url
      $url->scheme($param->{scheme}) if $param->{scheme};
      $url->port($param->{port}) if $param->{port};

      if ($param->{host}) {
        $url->host($param->{host});
        $url->port(undef) unless $param->{port};
        $url->scheme('http') unless $url->scheme;
      };

      # Clone query
      $url->query( [@{$param->{query}}] ) if $param->{query};

      # Get path
      my $path = $url->path;

      # Lookup match
      my $r = $match->root->find($param->{route});

      # Interpolate path
      my @parts;
      while ($r) {
        my $p = '';
        foreach my $part (@{$r->pattern->tree}) {
          my $t = $part->[0];

          # Slash
          if ($t eq 'slash') {
            $p .= '/';
          }

          # Text
          elsif ($t eq 'text') {
            $p .= $part->[1];
          }

          # Various wildcards
          elsif ($t =~ m/^(?:wildcard|placeholder|relaxed)$/) {
            if (exists $values{$part->[1]}) {
              $p .= $values{$part->[1]};
            }
            else {
              $p .= '{' . $part->[1] . '}';
            };
          };
        };

        # Prepend to path array
        unshift(@parts, $p);

        # Go up one level till root
        $r = $r->parent;
      };

      # Set path
      $path->parse(join('', @parts)) if @parts;

      # Fix trailing slash
      $path->trailing_slash(1)
        if (!$name || $name eq 'current')
        && $req->url->path->trailing_slash;

      # Make path absolute
      my $base_path = $base->path;
      unshift @{$path->parts}, @{$base_path->parts};
      $base_path->parts([]);

      # Interpolate url for query parameters
      return _interpolate($url->to_abs->to_string, \%values, $values);
    }
  );


  # Add 'get_endpoints' helper
  $mojo->helper(
    get_endpoints => sub {
      my $c = shift;

      # Get all endpoints
      my %endpoint_hash;
      foreach (keys %endpoints) {
        $endpoint_hash{$_} = $c->endpoint($_);
      };

      # Return endpoint hash
      return \%endpoint_hash;
    });
};


# Interpolate templates
sub _interpolate {
  my $endpoint = shift;

  # Decode escaped symbols
  $endpoint =~
    s/\%7[bB](.+?)\%7[dD]/'{' . b($1)->url_unescape . '}'/ge;

  my $param = shift;
  my $orig_param = shift;

  # Interpolate template
  pos($endpoint) = 0;
  while ($endpoint =~ /\{([^\}\?}\?]+)\??\}/g) {

    # Save search position
    # Todo: That's not exact!
    my $val = $1;
    my $p = pos($endpoint) - length($val) - 1;

    my $fill = undef;

    # Look in param
    if ($param->{$val}) {
      $fill = b($param->{$val})->url_escape;
      $endpoint =~ s/\{$val\??\}/$fill/;
    }

    # unset specific parameters
    elsif (exists $orig_param->{$val}) {

      # Delete specific parameters
      for ($endpoint) {
        if (s/(?<=[\&\?])[^\}][^=]*?=\{$val\??\}//g) {
          s/([\?\&])\&*/$1/g;
          s/\&$//g;
        };
        s/^([^\?]+?)([\/\.])\{$val\??\}\2/$1$2/g;
        s/^([^\?]+?)\{$val\??\}/$1/g;
      };
    };

    # Reset search position
    # Todo: (not exact if it was optional)
    pos($endpoint) = $p + length($fill || '');
  };

  # Ignore optional placeholders
  if (exists $param->{'?'} &&
        !defined $param->{'?'}) {
    for ($endpoint) {
      s/(?<=[\&\?])[^\}][^=]*?=\{[^\?\}]+?\?\}//g or last;
      s/([\?\&])\&*/$1/g;
      s/\&$//g;
    };
  };

  # Strip empty query marker
  $endpoint =~ s/\?$//;
  return $endpoint;
};


1;


__END__

=pod

=head1 NAME

Mojolicious::Plugin::Util::Endpoint - Use Template URIs in Mojolicious


=head1 SYNOPSIS

  # Mojolicious::Lite
  plugin 'Util::Endpoint';

  # Mojolicious
  $self->plugin('Util::Endpoint');

  my $rs = $mojo->routes;

  # Set endpoint
  my $r = $rs->route('/:user')->endpoint(
    webfinger => {
      query  => [
        q => '{uri}'
      ]
    });

  return $self->endpoint('webfinger');
  # https://sojolicio.us/{user}?q={uri}

  $self->stash(user => 'Akron');

  return $self->endpoint('webfinger');
  # https://sojolicio.us/Akron?q={uri}

  return $self->endpoint(webfinger => {
    uri => 'acct:akron@sojolicio.us'
  });
  # https://sojolicio.us/Akron?q=acct%3Aakron%40sojolicio.us


=head1 DESCRIPTION

L<Mojolicious::Plugin::Util::Endpoint> is a plugin that
allows for the simple establishment of endpoint URIs.
This is similar to L<url_for|Mojolicious::Controller/url_for>,
but includes support for template URIs with parameters
following L<RFC6570|https://tools.ietf.org/html/rfc6570> Level 1
(as used in, e.g., Host-Meta or OpenSearch).


=head1 METHODS

=head2 register

  # Mojolicious
  $app->plugin('Util::Endpoint');

  # Mojolicious::Lite
  plugin 'Util::Endpoint';

Called when registering the plugin.


=head1 SHORTCUTS

=head2 endpoint

  my $rs = $mojo->routes
  my $r = $rs->route('/suggest')->endpoint(
    opensearch => {
      scheme => 'https',
      host   => 'sojolicio.us',
      port   => 3000,
      query  => [
        q     => '{searchTerms}',
        start => '{startIndex?}'
      ]
    });

Establishes an endpoint defined for a service.
It accepts optional parameters C<scheme>, C<host>,
a C<port> and query parameters (C<query>),
overwriting the current values of C<url_for>.
Template parameters need curly brackets, optional
template parameters need a question mark before
the closing bracket.
Optional path placeholders are currenty not supported.
Returns the route.

B<Warning>: Support for named routes to use with
C<url_for> was dropped in v0.19.

=head1 HELPERS

=head2 endpoint

  # In Controller:
  #   Set endpoints:
  $self->endpoint(hub => 'http://sojolicio.us/search?q={searchTerm}');
  $self->endpoint(hub => Mojo::URL->new('http://pubsubhubbub.appspot.com/'));

  #   Get endpoints:
  return $self->endpoint('webfinger');
  return $self->endpoint(webfinger => { user => 'me' } );

  # Interpolate arbitrary template URIs
  return $self->endpoint(
    'http://sojolicio.us/.well-known/webfinger?resource={uri}&rel={rel?}' => {
      'uri' => 'acct:akron@sojolicio.us',
      '?'   => undef
    });

  # In Template:
  <%= endpoint 'webfinger' %>

Get or set endpoints defined for a specific service.

For setting it accepts the name of the endpoint and
either a string with the endpoint URI or a L<Mojo::URL> object.

For getting it accepts the name of the endpoint or an arbitrary
template URI and additional stash values for the route as a hash reference.
These stash values override existing stash values from
the controller and fill the template variables.

  # In Controller:
  return $self->endpoint('opensearch');
  # https://sojolicio.us/suggest?q={searchTerms}&start={startIndex?}

  return $self->endpoint(opensearch => {
    searchTerms => 'simpson',
    '?' => undef
  });
  # https://sojolicio.us/suggest?q=simpson

The special parameter C<?> can be set to C<undef> to ignore
all undefined optional template parameters.

If the defined endpoint can't be found, the value for C<url_for>
is returned.


=head2 get_endpoints

  # In Controller:
  my $hash = $self->get_endpoints;

  while (my ($key, $value) = each %$hash) {
    print $key, ' => ', $value, "\n";
  };

Returns a hash of all endpoints, interpolated with the current
controller stash.

B<Note:> This helper is EXPERIMENTAL and may be deprecated in further releases.


=head1 COMMANDS

=head2 endpoints

  $ perl app.pl endpoints

Show all endpoints of the app established by this plugin.


=head1 DEPENDENCIES

L<Mojolicious> (best with SSL support).


=head1 CONTRIBUTORS

L<Viacheslav Tykhanovskyi|https://github.com/vti>


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-Util-Endpoint


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011-2016, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the same terms as Perl.

=cut
